xls/dslx/frontend/ast.cc

```diff
@@ -401,28 +408,33 @@ absl::flat_hash_set<std::string> FreeVariables::Keys() const {
   return result;
 }
 
-FreeVariables GetFreeVariables(const AstNode* node, const Pos* start_pos) {
+FreeVariables GetFreeVariablesByLambda(
+    const AstNode* node,
+    const std::function<bool(const NameRef&)>& consider_free) {
   DfsIteratorNoTypes it(node);
   FreeVariables freevars;
   while (it.HasNext()) {
     const AstNode* n = it.Next();
     if (const auto* name_ref = dynamic_cast<const NameRef*>(n)) {
-      // If a start position was given we test whether the name definition
-      // occurs before that start position. (If none was given we accept all
-      // name refs.)
-      if (start_pos == nullptr) {
+      if (consider_free == nullptr || consider_free(*name_ref)) {
         freevars.Add(name_ref->identifier(), name_ref);
-      } else {
-        std::optional<Pos> name_def_start = name_ref->GetNameDefStart();
-        if (!name_def_start.has_value() || *name_def_start < *start_pos) {
-          freevars.Add(name_ref->identifier(), name_ref);
-        }
       }
     }
   }
   return freevars;
 }
 
+FreeVariables GetFreeVariablesByPos(const AstNode* node, const Pos* start_pos) {
+  std::function<bool(const NameRef&)> consider_free = nullptr;
+  if (start_pos != nullptr) {
+    consider_free = [start_pos](const NameRef& name_ref) {
+      std::optional<Pos> name_def_start = name_ref.GetNameDefStart();
+      return !name_def_start.has_value() || name_def_start.value() < *start_pos;
```

Comment:
I'm probably just missing something - but I don't understand this code. It seems like we're saying that we should consider the variable free if:

1. it's not defined anywhere (seems correct), or
2. the definition comes *before* the current position... which seems backwards, since that sounds like the situation in which the variable would be bound?

I acknowledge that this seems to work, and isn't a change from what was here before, but it seems worth a comment to clarify how.

Reply:
Did you see the comment in the header file?

```
// Retrieves all the free variables (references to names that are defined                                                               
// prior to start_pos) that are transitively in this AST subtree.                                                                       
//
// For example, if given the AST node for this function:                                                                                
//
//    const FOO = u32:42;
//    fn main(x: u32) { FOO+x }                                                                                                         
//
// And *using the starting point of the function* as the `start_pos`, the `FOO`                                                         
// will be flagged as a free variable and returned.
```

The `start_pos` given is a way to approximate "free variable with respect to this AST construct". i.e. all the references with defs that are defined before this `start_pos` point are considered free. This gives an easy way to say "everything defined inside the body we don't need to worry about, only tell me about references to things before this lexical position in the file.

LMK if you have a way you'd rephrase that header comment for more clarity.

Reply:
(For now just added this paragraph to the API comment as an additional note, on the assumption maybe it's rephrasing in a clearer way.)